feat: scaffolding ca certs

lkingland opened the pull request:
- :gift: Scaffolded functions include root CA certs
- :gift: adds make target for updating embedded certs

All functions which utilize scaffolding (Go for now, but will work for all) will now include root CA certs, as provided by the curl project.

matejvasek commented on generate/templates/main.go:
```diff
@@ -11,6 +11,8 @@ import (
 	"path/filepath"
 )
 
+// MODIFIED
```
?

matejvasek commented on Makefile:
```diff
@@ -86,137 +127,179 @@ clean_templates:
-generate/zz_filesystem_generated.go: clean_templates
+# the filesysetm file itself is defined as being fresh if all the files in
+# templates are fresh as well as the certificates file.
+generate/zz_filesystem_generated.go: $(TEMPLATES) templates/certs/ca-certificates.crt
```
I very much on purpose remove similar condition in the past. This can detect that file was modified but it won't for instance detect file deletion.

matejvasek replied:
This should be phony.

matejvasek replied:
The `generate/zz_filesystem_generated.go` recipe takes 200ms. Most of which is actually spend on `go run`'s on the fly compilation. The actual template generation takes 40ms. I don't see why we cannot run it all the time.

lkingland replied:
I have removed all but the minimal makefile updates and will submit them separately for discussion.

matejvasek commented on pkg/filesystem/filesystem_test.go:
```diff
@@ -117,6 +117,11 @@ func loadFS(fileSys filesystem.Filesystem) ([]FileInfo, error) {
+			if runtime.GOOS == "windows" {
+				// TODO I have no idea why, but symlinks are ever so slightly different
+				// in Windows and fail the test.
+				return nil
```
Please do not just ignore it. Use `filepath.ToSlash()` to fixup the value.

matejvasek replied:
Actually at this place it will be always slash separated since `fs.FS` uses just slashes.

matejvasek replied:
Just remove the `if` altogether.

matejvasek commented on pkg/filesystem/filesystem_test.go:
```diff
@@ -157,6 +162,11 @@ func loadLocalFiles(root string) ([]FileInfo, error) {
+			if runtime.GOOS == "windows" {
+				// TODO I have no idea why, but symlinks are ever so slightly different
+				// in Windows and fail the test.
+				return nil
```
Please do not just ignore it. Use `filepath.ToSlash()` to fixup the value.

matejvasek commented on pkg/filesystem/filesystem_test.go:
```diff
@@ -117,6 +117,11 @@ func loadFS(fileSys filesystem.Filesystem) ([]FileInfo, error) {
 				return err
 			}
 		case fs.ModeSymlink:
+			if runtime.GOOS == "windows" {
+				// TODO I have no idea why, but symlinks are ever so slightly different
+				// in Windows and fail the test.
+				return nil
+			}
```
```suggestion
```

matejvasek commented on pkg/filesystem/filesystem_test.go:
```diff
@@ -157,6 +162,11 @@ func loadLocalFiles(root string) ([]FileInfo, error) {
 				return err
 			}
 		case fs.ModeSymlink:
+			if runtime.GOOS == "windows" {
+				// TODO I have no idea why, but symlinks are ever so slightly different
+				// in Windows and fail the test.
+				return nil
+			}
 			t, _ := os.Readlink(path)
 			bs = []byte(t)
```
```suggestion
			t, _ := os.Readlink(path)
			bs = []byte(filepath.ToSlash(t))
```

matejvasek commented on Makefile:
```diff
@@ -86,137 +127,179 @@ clean_templates:
+
+.PHONY: certs-check
+certs-check: certs ## Check that root certificates are up-to-date
+	@if [[ -n $$(git diff -- schema/func_yaml-schema.json) ]]; \
```
How does this check certs?

lkingland replied:
Renamed this `certs-modified-check` and added better comments 👍🏻

matejvasek replied:
I was getting on `schema/func_yaml-schema.json`. Why it is there?

lkingland replied:
I have removed all but the minimal makefile updates and will submit them separately for discussion.

matejvasek commented on Makefile:
```diff
@@ -86,137 +127,179 @@ clean_templates:
-#############
+.PHONY: templates-check
+templates-check: certs ## Check that the generated embedded filesystems are up-to-date
+	@if [[ -n $$(git diff -- schema/func_yaml-schema.json) ]]; \
```
How does this check `generate/zz_filesystem_generated.go`?

lkingland replied:
Yes that comment is confusing.  I removed this one and marked the templates target as .PHONY such that it always builds, the other two instances (certs and schema) were renamed to `x-modified-check` and removed the comment such that they are "hidden" from `make help`.  These are used in source control along with a `make clean` to confirm that the files are up-to-date.

matejvasek replied:
I was getting on `schema/func_yaml-schema.json`. Why it is there?

lkingland replied:
I have removed all but the minimal makefile updates and will submit them as a separate PR for discussion

matejvasek commented:
On similar note like https://github.com/knative/func/pull/1823#discussion_r1238839860.
I really do not like using `$(CODE)` as prerequisites -- it cannot detect file deletion (only addtition/modification).
Moreover `go build`, `go test`, `golangci-lint` are smart enough to know when and what re-compilation/re-test is required.
Similarly I don't see much value in these stamps, what I see is possible risk of not running recipe when is should be run.
I rather run recipe that takes 500ms all the time  than risking that recipe is not done when it really needed.

lkingland commented:
Thanks for the review @matejvasek , I have added `filepath.ToSlash()` to the filesystem symlink test, and removed all but the minimum makefile updates for discussion in a separate PR.

matejvasek commented:
/approve
/lgtm

matejvasek commented:
The windows test are failing on the symlink despite `filepath.ToSlash()`, it makes no sense.

lkingland commented:
> The windows test are failing on the symlink despite `filepath.ToSlash()`, it makes no sense.

Shucks.  I was hoping the `ToSlash` would do the trick.  I really don't mind if they are a few bytes different, so long as they are both symlinks.

matejvasek commented:
> > The windows test are failing on the symlink despite `filepath.ToSlash()`, it makes no sense.
> 
> Shucks. I was hoping the `ToSlash` would do the trick. I really don't mind if they are a few bytes different, so long as they are both symlinks.

@lkingland I think I know where the issue might be. The unit test in GH also regenerates zz file on Windows. The zip generate script should also fixup links.

matejvasek commented:
>actual app do use zz file generated on Linux 

Actually I am not 100% positive about this. I am only sure that we version file generated on Linux.

matejvasek commented:
@lkingland  tl;dr we should not suppress the tests, it really indicates that something is wrong

We need `filepath.ToSlash()` in:
* [zip generator](https://github.com/knative/func/blob/main/generate/templates/main.go#L83C4-L83C43)
* [os fs](https://github.com/knative/func/blob/main/pkg/filesystem/filesystem.go#L176)
* [fs test](https://github.com/knative/func/blob/main/pkg/filesystem/filesystem_test.go#L161)

I know this is bug in existing code, not it this PR, sill we should fix it.

matejvasek commented:
/approve
/lgtm
